Sources/Components/CarouselView/CarouselView.swift

```diff
@@ -0,0 +1,192 @@
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: - Overrides
+
+    public override var backgroundColor: UIColor? {
+        didSet {
+            collectionView.backgroundColor = backgroundColor
+        }
+    }
+
+    public override func layoutSubviews() {
+        super.layoutSubviews()
+
+        guard initialLayout else { return }
+        initialLayout = false
+
+        // Scroll to first item
+        collectionView.contentOffset = CGPoint(
+            x: bounds.width,
+            y: collectionView.contentOffset.y
+        )
+    }
+}
+
+public extension CarouselView {
+    var contentOffset: CGPoint {
+        return CGPoint(
+            x: collectionView.contentOffset.x - collectionView.bounds.width,
+            y: collectionView.contentOffset.y
+        )
+    }
+
+    func register(_ cellClass: CarouselViewCell.Type) {
+        collectionView.register(cellClass)
+    }
+
+    func dequeue<T>(_ cellType: T.Type, for indexPath: IndexPath) -> T where T: CarouselViewCell {
+        return collectionView.dequeue(cellType, for: indexPath)
+    }
+
+    func scrollToItem(at indexPath: IndexPath, at scrollPosition: UICollectionView.ScrollPosition, animated: Bool) {
+        collectionView.scrollToItem(at: IndexPath(item: indexPath.item + 1, section: 0), at: scrollPosition, animated: animated)
+    }
+}
+
+extension CarouselView: UICollectionViewDataSource {
+    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        guard let dataSource = dataSource else {
+            return 0
+        }
+
+        return dataSource.numberOfItems(in: self) + 2
+    }
+
+    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let dataSource = dataSource else {
+            preconditionFailure("Data source is not available")
+        }
+
+        return dataSource.carouselView(self, cellForItemAt: self.indexPath(forItem: indexPath.item))
+    }
+}
+
+extension CarouselView: UICollectionViewDelegateFlowLayout {
+    public func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return bounds.size
+    }
+
+    public func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        delegate?.carouselView(self, didSelectItemAt: self.indexPath(forItem: indexPath.item))
+    }
+
+    public func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
+        delegate?.carouselViewDidEndDecelerating(self)
+    }
+
+    public func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint, targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        let numberOfItems = collectionView.numberOfItems(inSection: 0)
+
+        if scrollView.isPagingEnabled {
```

Comment:
We could use `guard` here for early return.

Reply:
You are right